orttraining/orttraining/core/graph/pipeline_transformer.cc

```diff
@@ -0,0 +1,255 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "orttraining/core/graph/pipeline_transformer.h"
+
+using namespace onnxruntime::common;
+
+namespace onnxruntime {
+namespace training {
+
+void GetPipelineSendOutput(const Graph& graph, std::string& loss_name) {
+  for (auto& node : graph.Nodes()) {
+    if (!node.OpType().compare("Send")) {
+      // send op should always have an output, which is the OutputSignal.
+      loss_name = node.OutputDefs()[0]->Name();
+      return;
+    }
+  }
+}
+
+bool IsBackward(Node& node) {
+  return (node.Description() == "Backward pass");
+}
+
+void AddInputEvent(Graph& graph, const std::string& op_name,
```

Comment:
I am wondering why it doesn't return the event (or its name) it adds?

Reply:
I'm ok either way. I feel this style saves one std::move for the returning string. But if you have a strong prefernce over the other, I can change it too.

---
In reply to: [409789701](https://github.com/microsoft/onnxruntime/pull/3513#discussion_r409789701) [](ancestors = 409789701)